libbpf-tools/fsdist.c

```diff
@@ -0,0 +1,489 @@
+// SPDX-License-Identifier: (LGPL-2.1 OR BSD-2-Clause)
+// Copyright (c) 2021 Hengqi Chen
+//
+// 20-May-2021   Hengqi Chen   Created this.
+#include <argp.h>
+#include <signal.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <time.h>
+#include <unistd.h>
+
+#include <bpf/libbpf.h>
+#include <bpf/bpf.h>
+
+#include "fsdist.h"
+#include "fsdist.skel.h"
+#include "trace_helpers.h"
+
+#define warn(...) fprintf(stderr, __VA_ARGS__)
+
+enum fs_type {
+	BTRFS,
+	EXT4,
+	NFS,
+	XFS,
+};
+
+static struct fs_config {
+	const char *fs;
+	const char *op_funcs[MAX_OP];
+} fs_configs[] = {
+	{ "btrfs", {
+		[READ] = "btrfs_file_read_iter",
+		[WRITE] = "btrfs_file_write_iter",
+		[OPEN] = "btrfs_file_open",
+		[FSYNC] = "btrfs_sync_file",
+		[GETATTR] = NULL, /* not supported */
+	}},
+	{ "ext4", {
+		[READ] = "ext4_file_read_iter",
+		[WRITE] = "ext4_file_write_iter",
+		[OPEN] = "ext4_file_open",
+		[FSYNC] = "ext4_sync_file",
+		[GETATTR] = "ext4_file_getattr",
+	}},
+	{ "nfs", {
+		[READ] = "nfs_file_read",
+		[WRITE] = "nfs_file_write",
+		[OPEN] = "nfs_file_open",
+		[FSYNC] = "nfs_file_fsync",
+		[GETATTR] = "nfs_getattr",
+	}},
+	{ "xfs", {
+		[READ] = "xfs_file_read_iter",
+		[WRITE] = "xfs_file_write_iter",
+		[OPEN] = "xfs_file_open",
+		[FSYNC] = "xfs_file_fsync",
+		[GETATTR] = NULL, /* not supported */
+	}},
+};
+
+static char *file_op_names[] = {
+	[READ] = "read",
+	[WRITE] = "write",
+	[OPEN] = "open",
+	[FSYNC] = "fsync",
+	[GETATTR] = "getattr",
+};
+
+static struct hist zero;
+static volatile sig_atomic_t exiting;
+
+// options
+static enum fs_type fs_type = EXT4;
+static bool emit_timestamp = false;
+static bool timestamp_in_ms = false;
+static pid_t target_pid = 0;
+static int interval = 99999999;
+static int count = 99999999;
+static bool verbose = false;
+
+const char *argp_program_version = "fsdist 0.1";
+const char *argp_program_bug_address =
+	"https://github.com/iovisor/bcc/tree/master/libbpf-tools";
+const char argp_program_doc[] =
+"Summarize btrfs/ext4/nfs/xfs operation latency.\n"
```

Comment:
"Summarize file system operations latency." to keep it short and generic?

Then we can have a list of supported file systems below.